src/hasSubObject.test.ts

```diff
@@ -0,0 +1,158 @@
+    expect(hasSubObject({ a: 2, b: "b", c: 3 }, { a: 1, b: "b" })).toBe(false);
+  });
+
+  test("works with deep objects", () => {
+    expect(hasSubObject({ a: { b: 1, c: 2 } }, { a: { b: 1, c: 2 } })).toBe(
+      true,
+    );
+    expect(hasSubObject({ a: { b: 1, c: 2 } }, { a: { b: 1, c: 0 } })).toBe(
+      false,
+    );
+  });
+});
+
+describe("data last", () => {
+  test("empty sub-object", () => {
+    expect(pipe({ a: 1, b: 2, c: 3 }, hasSubObject({}))).toBe(true);
+    expect(pipe({}, hasSubObject({}))).toBe(true);
+  });
+
+  test("works with primitives", () => {
+    expect(pipe({ a: 1, b: "b", c: 3 }, hasSubObject({ a: 1, b: "b" }))).toBe(
+      true,
+    );
+    expect(pipe({ a: 1, b: "c", c: 3 }, hasSubObject({ a: 1, b: "b" }))).toBe(
+      false,
+    );
+    expect(pipe({ a: 2, b: "b", c: 3 }, hasSubObject({ a: 1, b: "b" }))).toBe(
+      false,
+    );
+  });
+
+  test("works with deep objects", () => {
+    expect(
+      pipe({ a: { b: 1, c: 2 } }, hasSubObject({ a: { b: 1, c: 2 } })),
+    ).toBe(true);
+
+    expect(pipe({ a: { b: 1, c: 2 } }, hasSubObject({ a: { b: 1 } }))).toBe(
+      false,
+    );
+  });
+});
+
+describe("typing", () => {
+  describe("data-first", () => {
+    test("must have matching keys and values", () => {
+      expectTypeOf(hasSubObject({ a: 2 }, { a: 1 })).toEqualTypeOf<boolean>();
+
+      // @ts-expect-error - missing a key
+      hasSubObject({ b: 2 }, { a: 1 });
+
+      // @ts-expect-error - different value type
+      hasSubObject({ a: "a" }, { a: 1 });
+
+      // ok - wider value type
+      hasSubObject({ a: "a" } as { a: number | string }, { a: 1 });
+
+      // @ts-expect-error - narrower value type
+      hasSubObject({ a: "a" }, { a: 1 } as { a: number | string });
+
+      // ok - unknown data type
+      expectTypeOf(
+        hasSubObject({ a: 2 } as unknown, { a: 1 }),
+      ).toEqualTypeOf<boolean>();
+
+      // ok - const value
+      expectTypeOf(
+        hasSubObject({ a: 2 }, { a: 1 } as const),
+      ).toEqualTypeOf<boolean>();
+    });
+
+    test("allows nested objects", () => {
+      // ok - nested object has extra keys
+      expectTypeOf(
+        hasSubObject({ a: { b: 4 } }, { a: { b: 1, c: 2 } }),
+      ).toEqualTypeOf<boolean>();
+
+      // @ts-expect-error - nested object has missing keys
+      hasSubObject({ a: { b: 1, c: 2 } }, { a: { b: 1 } });
+
+      // @ts-expect-error - nested object has wrong value types
+      hasSubObject({ a: { b: 4, c: "c" } }, { a: { b: 1, c: 2 } });
+    });
+
+    test("narrows type", () => {
+      const obj = {} as { a?: string; b?: number };
+
+      if (hasSubObject(obj, { a: "a" })) {
+        expectTypeOf(obj).toEqualTypeOf<{ a: string; b?: number }>();
+      }
+
+      if (hasSubObject(obj, { a: "a" } as const)) {
+        expectTypeOf(obj).toEqualTypeOf<{ a: "a"; b?: number }>();
+      }
+    });
+  });
+
+  describe("data-last", () => {
```

Comment:
Generally speaking about testing dataLast, I prefer testing everything via a pipe because the inference works a little differently, and it represents real-life usage a little better. Can you add a pipe to all the dataLast invocations?